Update performance-monitoring.mdx (#13001)
getsentry/sentry-docs: docs/product/sentry-basics/performance-monitoring.mdx

---
title: "Performance Monitoring V2"
sidebar_order: 1
description: "Understand and monitor how your application performs in production. Track key metrics, analyze bottlenecks, and resolve performance issues with distributed tracing, detailed transaction data, and automated issue detection."
---

<Include name="performance-moving.mdx" />

> In many tools, Performance Monitoring is just about tracking a few key metrics on your web pages. Sentry takes a different approach. By setting up [Distributed Tracing](/product/sentry-basics/tracing/distributed-tracing/), Sentry captures detailed performance data for [every transaction](/concepts/key-terms/tracing/) in your entire application stack and automatically presents it in a variety of easy-to-use but powerful features so you can rapidly identify and resolve performance issues as they happen - all in one place.

Reviewer: I got a little lost clicking on the link for 'performance data for every transaction' and then being routed to the 'tracing' key terms page. Am I missing something?

Author: hmm, let's say "captures detailed performance data for every transaction [(trace)](link)

Author: resolevd in commit

> In many tools, Performance Monitoring is just about tracking a few key metrics on your web pages. Sentry takes a different approach. By setting up [Distributed Tracing](/product/sentry-basics/tracing/distributed-tracing/), Sentry captures detailed performance data for [every transaction](/concepts/key-terms/tracing/) in your entire application stack and automatically presents it in a variety of easy-to-use but powerful features so you can rapidly identify and resolve performance issues as they happen - all in one place.

Reviewer: ```suggestion
In many tools, Performance Monitoring is just about tracking a few key metrics on your web pages; Sentry takes a more granular approach. By setting up [Distributed Tracing](/product/sentry-basics/tracing/distributed-tracing/), Sentry captures detailed performance data for [every transaction](/concepts/key-terms/tracing/) in your entire application stack and automatically presents it in a variety of easy-to-use but powerful features so you can rapidly identify and resolve performance issues as they happen - all in one place.
```

<Alert>

> Set up [Distributed Tracing](/product/sentry-basics/tracing/) to enable Performance Monitoring.

Reviewer: This navigates to the same link as the 'every transaction' link above, I would avoid this as search crawlers may see it as link stuffing.

Author: good point let's remove it

Author: actually, let's remove the one above. Would rather have the link here in the alert.

Author: deciding to keep this one

</Alert>

Sentry has multiple tools for monitoring performance which are useful for different aspects of your monitoring & debugging needs:

> ### 1: Insights: Opinionated views of performance data for each layer of your stack

Reviewer: ```suggestion
### 1: Insights: Opinionated Views of Performance Data for Each Layer of Your Stack
```

> After turning on Distributed Tracing, you'll automatically see a set of [Insights](/product/insights/) pages that give you a high-level view of performance data for each layer of your stack. This is extremely useful for quickly seeing any major performance outliers that need optimization, or tracking potential regressions across time, environments, releases, etc.

Reviewer: ```suggestion
After turning on Distributed Tracing, you'll automatically see a set of [Insights](/product/insights/) pages that give you a high-level view of performance data for each layer of your stack. This is useful for quickly seeing any major performance outliers that need optimization, or tracking potential regressions across time, environments, releases, etc.
```

> - [**Frontend Insights**](/product/insights/frontend/) give you a view of performance data for your frontend (browser) transactions. These are broken down into pageview performance with things like Core Web Vitals, as well as transaction performance with things like duration, and error rate.

Reviewer: Are frontend insights always browser transactions? Couldn't they also be native app transactions or something different like game interactions? Would it be worth updating this to something like '(typically browser transactions)' or something?

Author: resolved in commit

> - [**Backend Insights**](/product/insights/backend/) show your backend transaction performance, across things like database queries, external API calls, caches, queues, crons, and more.

Reviewer: ```suggestion
- [**Backend Insights**](/product/insights/backend/) show your backend transaction performance across things like database queries, external API calls, caches, queues, crons, and more.
```

- [**Mobile Insights**](/product/insights/mobile/) track vital performance metrics for your mobile apps, including your cold & warm app start times, TTID & TTFD, and time to load or render key screens throughout your app. 
- [**AI Insights**](/product/insights/ai/) give you a view of your LLM performance. This is currently in beta and only available for certain integrations.

All accounts have access to the top level insights pages. Upgrading to the [Team plan](https://sentry.io/signup/?original_referrer=https%3A%2F%2Fwww.google.com%2F&period=annual&plan=am1_team_auf&trial) or higher gets you access to more dedicated insights pages for common monitoring use cases.

> <Arcade src="https://demo.arcade.software/BTvUjriSDjaHiVqPvIFP?embed&embed_mobile=tab&embed_desktop=inline&show_copy_link=true" /><br />

Reviewer: I'd recommend adding a little context snippet to get people to engage with this Arcade. Maybe along the lines of: 'See the video below for a real-time walkthrough of performance insights:'

Author: resolved in commit

> ### 2: Performance Issues: Automatically identify full-stack bottlenecks

Reviewer: ```suggestion
### 2: Performance Issues: Automatically Identify Full-Stack Bottlenecks
```

> Alongside typical [error issues](/product/issues/issue-details/error-issues/), Sentry will automatically detect common performance issues and group them into [performance issues](/product/issues/issue-details/performance-issues/). For instance: 

Reviewer: ```suggestion
Just like typical [error issues](/product/issues/issue-details/error-issues/), Sentry will automatically detect common performance issues and group them into [performance issues](/product/issues/issue-details/performance-issues/). For instance: 
```

- N+1 Query & API call issues that indicate unnecessary round-trips to your backend
- Consecutive DB Queries or HTTP Requests that could be batched for better performance
- Database queries or File I/O running on the main thread, which can block the UI thread and cause performance issues
- And many more documented [here](/product/issues/performance-issues/). 

<Arcade src="https://demo.arcade.software/t8kZ9OeigLNrH3fLynIu?embed&embed_mobile=tab&embed_desktop=inline&show_copy_link=true" /> <br />

> ### 3: Trace Explorer: Find root causes & common patterns for performance issues

Reviewer: ```suggestion
### 3: Trace Explorer: Find Root Causes & Common Patterns for Performance Issues
```

> While Performance Issues and Insights give you a high-level view of performance data, the [Trace Explorer](/product/issues/performance-issues/trace-explorer/) lets you slice and dice all your performance data by any attribute. This is an extremely powerful tool for finding root causes and common patterns for performance regressions or potential optimizations. For instance, you can find the p95 duration of all spans on the `GET /users/:id` endpoint, or find all pages slowing down because of a particular middleware routing. 

Reviewer: This 'Trace Explorer' link is a 404.

Author: resolved in commit (cursor hallucinated this link rip)

> While Performance Issues and Insights give you a high-level view of performance data, the [Trace Explorer](/product/issues/performance-issues/trace-explorer/) lets you slice and dice all your performance data by any attribute. This is an extremely powerful tool for finding root causes and common patterns for performance regressions or potential optimizations. For instance, you can find the p95 duration of all spans on the `GET /users/:id` endpoint, or find all pages slowing down because of a particular middleware routing. 

Reviewer: ```suggestion
While Performance Issues and Insights give you a high-level view of performance data, the [Trace Explorer](/product/issues/performance-issues/trace-explorer/) lets you slice and dice all your performance data by any attribute. This is a powerful tool for finding root causes and common patterns for performance regressions or potential optimizations. For instance, you can find the p95 duration of all spans on the `GET /users/:id` endpoint, or find all pages slowing down because of a particular middleware routing. 
```

> The newest version of the Trace Explorer also lets you quickly calculate metrics based on your span data, like p50, p95, and p99 durations, and then group them by any attribute. This lets you quickly see if there's specific endpoints, versions, user groups, etc. that are performing poorly. For a more thorough dive into how you can use the new Trace Explorer and Span Metrics, take a look at our most recent [blog post](https://blog.sentry.io/find-and-fix-performance-bottlenecks-with-sentrys-trace-explorer/). 

Reviewer: ```suggestion
The newest version of the Trace Explorer also lets you quickly calculate metrics based on your span data, like p50, p95, and p99 durations, and then group them by any attribute. This lets you quickly see if there are specific endpoints, versions, user groups, etc. that are performing poorly. For a more thorough dive into how you can use the new Trace Explorer and Span Metrics, take a look at our most recent [blog post](https://blog.sentry.io/find-and-fix-performance-bottlenecks-with-sentrys-trace-explorer/). 
```

<Alert>

This demo shows a [new version](/product/explore/new-trace-explorer/) of the Trace Explorer that is available for all self-serve orgs in the Early Adopter program. Enable the EA flag in your [organization settings](https://sentry.io/orgredirect/settings/:orgslug/) to gain access to Span Metrics, Group By, Create Widgets and Alerts. 

</Alert>

<Arcade src="https://demo.arcade.software/E04YSJpq1w8bpk18Q5Kp" /> <br />

> ### 4: Metrics, Alerts, and Dashboards: Build custom views of performance data

Reviewer: ```suggestion
### 4: Metrics, Alerts, and Dashboards: Build Custom Views of Performance Data
```

By default, Sentry automatically tracks a set of [performance metrics](/platforms/javascript/tracing/span-metrics/performance-metrics/) for every trace. These are used to pupulate the Insights dashboards, but you can also use them to build custom [dashboards](/product/dashboards/) and [alerts](/product/alerts/).

> Opting into the [Early Adopter](https://sentry.io/orgredirect/settings/:orgslug/) program for the Trace Explorer also gives you access to the new [Span Metrics](/product/explore/new-trace-explorer/span-metrics/) feature, which lets you calculate custom metrics based on your span data. This is extremely useful for building dashboards and alerts that are underpinned by real performance tracing data, so you can quickly debug performance issues. 

Reviewer: Span Metrics link here is a 404.

Author: missed the # for a heading deeplink, resolved

> Opting into the [Early Adopter](https://sentry.io/orgredirect/settings/:orgslug/) program for the Trace Explorer also gives you access to the new [Span Metrics](/product/explore/new-trace-explorer/span-metrics/) feature, which lets you calculate custom metrics based on your span data. This is extremely useful for building dashboards and alerts that are underpinned by real performance tracing data, so you can quickly debug performance issues. 

Reviewer: ```suggestion
Opting into the [Early Adopter](https://sentry.io/orgredirect/settings/:orgslug/) program for the Trace Explorer also gives you access to the new [Span Metrics](/product/explore/new-trace-explorer/span-metrics/) feature, which lets you calculate custom metrics based on your span data. This is useful for building dashboards and alerts that are underpinned by real performance tracing data, so you can quickly debug performance issues. 
```

> ## How to set up Performance Monitoring

Reviewer: ```suggestion
## How To Set Up Performance Monitoring
```

For basic frontend or backend performance monitoring, just turn on Tracing in your relevant SDK. This will automatically trace every network request and give you a basic set of performance metrics. However, to make sure Sentry is capturing accurate and complete performance data, you need to set up [Distributed Tracing](/product/sentry-basics/distributed-tracing/). NextJS, Laravel, Ruby on Rails and a few others have strong out-of-the-box support for distributed tracing, but for more complex or multi-language applications, you'll need to configure this manually. Learn more about trace propagation [here](/platforms/javascript/tracing/distributed-tracing/). 

> ## How Performance Monitoring works under the hood

Reviewer: ```suggestion
## How Performance Monitoring Works Under the Hood
```

> Instead of instrumenting pageloads with specific metrics at set points, ie. `onLoad`, `onUnload`, `onError`, etc., Sentry automatically traces every network request, file I/O, and other operation in your application. This lets you see the performance impact of every request, and any errors that occur during that request. 

Reviewer: ```suggestion
Instead of instrumenting pageloads with specific metrics at set points, like `onLoad`, `onUnload`, `onError`, etc., Sentry automatically traces every network request, file I/O, and other operation in your application. This lets you see the performance impact of every request, and any errors that occured during that request. 
```
